### ISSUE TITLE
Mode executor with multiple custom modes

### ISSUE THREAD

**Issue by Alefunk** [Auterion/px4-ros2-interface-lib#65]
Hello, thanks for the great work!
I was wondering how to implement a mode executor with multiple custom modes. It's clear to me that one executor just owns one mode and thus is only capable of accessing the mode ID of the owned mode. However, is there a specific way to handle switching to other custom modes with unknown mode IDs during execution or is it not intended to use multiple custom modes within one executor?

**Comment by bkueng** [comment]
Hi, thanks!
If you instantiate both modes in the same binary, you can switch to the other mode by its [id](https://github.com/Auterion/px4-ros2-interface-lib/blob/main/px4_ros2_cpp/include/px4_ros2/components/mode.hpp#L150).
And apart than that, my idea was to add another interface that allows to switch to a mode by name, but that also requires changes on the PX4 side.

**Comment by Alefunk** [comment]
Thanks for the quick response. I actually made a mistake in my thinking. I solved it now by creating my own version of the [NodeWithModeExecutor](https://github.com/Auterion/px4-ros2-interface-lib/blob/main/px4_ros2_cpp/include/px4_ros2/components/node_with_mode.hpp#L97) and adding my custom modes in a std::map, which takes the [State](https://github.com/Auterion/px4-ros2-interface-lib/blob/main/examples/cpp/modes/mode_with_executor/include/mode.hpp#L66) enum from the ModeExecutor and the corresponding additional modes as input. This map is then passed to the constructor of the ModeExecutor in addition to its owned mode to access the respective ids.

**Comment by Alefunk** [comment]
Unfortunately, I have to reopen the issue. It seems like the first mode being registered is not available in PX4. I registered three modes and one executor owning a mode in various sequences. According to the debug console output all modes are successfully registered:

```
[control_node-13] [DEBUG] [1738755096.738561734] [control]: Waiting for FMU...
[control_node-13] [DEBUG] [1738755099.837483942] [control]: Checking message compatibility...
[control_node-13] [DEBUG] [1738755099.846124583] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755099.946666106] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.245420842] [control]: Registering 'Mode1' (arming check: 1, mode: 1, mode executor: 1)
[control_node-13] [DEBUG] [1738755100.245548324] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755100.345802521] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.345987783] [control]: Got RegisterExtComponentReply
[control_node-13] [DEBUG] [1738755100.346201681] [control]: Waiting for FMU...
[control_node-13] [DEBUG] [1738755100.359314316] [control]: Checking message compatibility...
[control_node-13] [DEBUG] [1738755100.362531787] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755100.362782242] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.654413544] [control]: Registering 'Mode2' (arming check: 1, mode: 1, mode executor: 0)
[control_node-13] [DEBUG] [1738755100.654496747] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755100.654550798] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.673450565] [control]: Got RegisterExtComponentReply
[control_node-13] [DEBUG] [1738755100.673810832] [control]: Waiting for FMU...
[control_node-13] [DEBUG] [1738755100.687244311] [control]: Checking message compatibility...
[control_node-13] [DEBUG] [1738755100.690082380] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755100.690186441] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.984618216] [control]: Registering 'Mode3' (arming check: 1, mode: 1, mode executor: 0)
[control_node-13] [DEBUG] [1738755100.984706709] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755100.984777544] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755100.990396635] [control]: Got RegisterExtComponentReply
[control_node-13] [DEBUG] [1738755100.990596891] [control]: Waiting for FMU...
[control_node-13] [DEBUG] [1738755101.003842911] [control]: Checking message compatibility...
[control_node-13] [DEBUG] [1738755101.006948154] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755101.007057168] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755101.308170770] [control]: Registering 'Mode4' (arming check: 1, mode: 1, mode executor: 0)
[control_node-13] [DEBUG] [1738755101.309669502] [control]: Subscriber found, continuing
[control_node-13] [DEBUG] [1738755101.309829017] [control]: Publisher found, continuing
[control_node-13] [DEBUG] [1738755101.323910396] [control]: Got RegisterExtComponentReply
[control_node-13] [DEBUG] [1738755101.328921943] [control]: Arming check request (id=4, only printed once)
```

When I check the status of the commander in the QGC Mavlink console as suggested in the documentation, I can only see the last three modes though:

```
INFO  [commander] Armed
INFO  [commander] navigation mode: Hold
INFO  [commander] user intended navigation mode: Hold
INFO  [commander] in failsafe: no
INFO  [commander] External Mode 2: nav_state: 24, name: Mode2
INFO  [commander] External Mode 3: nav_state: 25, name: Mode3
INFO  [commander] External Mode 4: nav_state: 26, name: Mode4
commander: cycle: 2835 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
commander: preflight check: 319 events, 0us elapsed, 0.00us avg, min 0us max 0us 0.000us rms
```

The first registered mode is always omitted, independently from it being an additional mode or the owned mode of the executor.

**Comment by bkueng** [comment]
Is there any error message on the PX4 console? Do you test in SITL?
I can try to reproduce.

**Comment by Alefunk** [comment]
Yes, I test in SITL with PX4 1.15.2 and am on main of px4-ros2-interface-lib with ROS Humble. There seem to be no error messages in the console except denying my mode change to the non existing mode with "notify negative". That would be great!

**Comment by Alefunk** [comment]
After some more fiddling around, the problem also seems related to the order of construction and registration of the modes. In my example, the modes are first constructed and then registered. If I construct and register each mode after each other, the behavior changes. Not the first registered mode is missing but another one. And the behavior also seems non-deterministic. Sometimes multiple modes are not registered depending on the order of registration.

**Comment by bkueng** [comment]
Hmm I tried with your branch and restarted your example at least 10 times, and the 3 modes with executor always appear.

One minor difference is that I'm using [this PX4 branch](https://github.com/PX4/PX4-Autopilot/pull/24113), which is basically `main`. Did you also try with PX4 `main`?

> On a sidenote, I run into an issue with pre-commit: ament_uncrustify will remove the parentheses, which are part of the [fold expression](https://github.com/Alefunk/px4-ros2-interface-lib/blob/a016150c3fac24760b7383d4bab0a78daa1554ae/px4_ros2_cpp/include/px4_ros2/components/node_with_mode.hpp#L122).

You'll need to add an ignore then. I generally prefer clang-format, but uncrustify is what ROS is using.

**Comment by Alefunk** [comment]
As you have suggested, we tried it on a different computer with PX4 main (gazebo version dependency changes) and there it works as expected. Might this behavior caused by an incompatible version of px4_msgs? I have checked out release/1.15 for compatibility to PX4 v1.15.2 and do not receive any incompatibility error messages from the px4-ros2-interface-lib, which is on main.

**Comment by bkueng** [comment]
I'm seeing it with 1.15.2 as well. It's [this commit](https://github.com/PX4/PX4-Autopilot/commit/4883f2128a02286df603d27104ed8a5110ee6d4c) that fixes it, which allows modes a bit more time.

**Comment by Alefunk** [comment]
I cherry-picked the commit into v.1.15.x and that solved the problem! Thank you very much for your support 👍 
If you are interested, I can create a merge request for the executor with multiple modes after some refining.

**Comment by bkueng** [comment]
Yes, certainly
